base/random/generation.jl

```diff
@@ -0,0 +1,321 @@
+               (Int64, UInt64), (Int128, UInt128), (Bool, UInt32)]
+
+    @eval RangeGenerator(r::UnitRange{$T}) = begin
+        isempty(r) && throw(ArgumentError("range must be non-empty"))
+        # overflow ok:
+        RangeGeneratorInt(first(r), convert($U, unsigned(last(r) - first(r)) + one($U)))
+    end
+end
+
+### RangeGenerator for BigInt
+
+struct RangeGeneratorBigInt <: RangeGenerator
+    a::BigInt         # first
+    m::BigInt         # range length - 1
+    nlimbs::Int       # number of limbs in generated BigInt's (z ∈ [0, m])
+    nlimbsmax::Int    # max number of limbs for z+a
+    mask::Limb        # applied to the highest limb
+end
+
+
+function RangeGenerator(r::UnitRange{BigInt})
+    m = last(r) - first(r)
+    m < 0 && throw(ArgumentError("range must be non-empty"))
+    nd = ndigits(m, 2)
+    nlimbs, highbits = divrem(nd, 8*sizeof(Limb))
+    highbits > 0 && (nlimbs += 1)
+    mask = highbits == 0 ? ~zero(Limb) : one(Limb)<<highbits - one(Limb)
+    nlimbsmax = max(nlimbs, abs(last(r).size), abs(first(r).size))
+    return RangeGeneratorBigInt(first(r), m, nlimbs, nlimbsmax, mask)
+end
+
+### rand(::RangeGenerator)
+
+# this function uses 32 bit entropy for small ranges of length <= typemax(UInt32) + 1
+# RangeGeneratorInt is responsible for providing the right value of k
+function rand(rng::AbstractRNG, g::RangeGeneratorInt{T,UInt64}) where T<:Union{UInt64,Int64}
+    local x::UInt64
+    if (g.k - 1) >> 32 == 0
+        x = rand(rng, UInt32)
+        while x > g.u
+            x = rand(rng, UInt32)
+        end
+    else
+        x = rand(rng, UInt64)
+        while x > g.u
+            x = rand(rng, UInt64)
+        end
+    end
+    return reinterpret(T, reinterpret(UInt64, g.a) + rem_knuth(x, g.k))
+end
+
+function rand(rng::AbstractRNG, g::RangeGeneratorInt{T,U}) where {T<:Integer,U<:Unsigned}
+    x = rand(rng, U)
+    while x > g.u
+        x = rand(rng, U)
+    end
+    (unsigned(g.a) + rem_knuth(x, g.k)) % T
+end
+
+function rand(rng::AbstractRNG, g::RangeGeneratorBigInt)
+    x = MPZ.realloc2(g.nlimbsmax*8*sizeof(Limb))
+    limbs = unsafe_wrap(Array, x.d, g.nlimbs)
+    while true
+        rand!(rng, limbs)
+        @inbounds limbs[end] &= g.mask
+        MPZ.mpn_cmp(x, g.m, g.nlimbs) <= 0 && break
+    end
+    # adjust x.size (normally done by mpz_limbs_finish, in GMP version >= 6)
+    x.size = g.nlimbs
+    while x.size > 0
+        @inbounds limbs[x.size] != 0 && break
+        x.size -= 1
+    end
+    MPZ.add!(x, g.a)
+end
+
+#### arrays
+
+function rand!(rng::AbstractRNG, A::AbstractArray, g::RangeGenerator)
+    for i in eachindex(A)
+        @inbounds A[i] = rand(rng, g)
+    end
+    return A
+end
+
+### rand(::UnitRange)
+
+rand(rng::AbstractRNG, r::UnitRange{<:Union{Signed,Unsigned,BigInt,Bool}}) =
+    rand(rng, RangeGenerator(r))
+
+rand!(rng::AbstractRNG, A::AbstractArray,
+      r::UnitRange{<:Union{Signed,Unsigned,BigInt,Bool,Char}}) =
```

Comment:
AFAICT, `UnitRange{Char}` cannot be constructed so this `Char` can be removed from this `Union`, which was on my todo list already.
This is a leftover from an old commit following `Char` not being an `Integer` anymore (cf. #13600).